### PR TITLE
Basics of tropical linear spaces

### PR DESCRIPTION
**Pull request by lkastner**
This branch contains the work on tropical linear spaces from the March22 KL workshop.

### CODE REVIEW

**Comment by fieker** [comment]
should this be merged? It is still somewhere WIP?

**Comment by lkastner** [comment]
Why was this merged? I specifically assigned to people to review it, it is their stuff. They have not done any reviewing yet. Also I am quite sure this was a draft.

In fact I have mentioned this PR on slack, I have talked to @VictoriaSchleis personally at the retreat. This means that both @YueRen and @VictoriaSchleis were aware of their duty to review.

**Comment by lkastner** [comment]
Ok, apparently this was discussed offline, so everything is fine.

File: src/TropicalGeometry/hypersurface.jl [hunk 1]
```diff
@@ -253,3 +253,14 @@ function polynomial(TH::TropicalHypersurface{M,EMB}) where {M,EMB}
...
 end
+export polynomial
+
+
+@doc Markdown.doc"""
+    minimalPolynomialFromHypersurface(T::TropicalHypersurface)
+
+Returns the minimal polynomial with smallest possible coefficients of a hypersurface
+"""
+function minimalPolynomialFromHypersurface(T::TropicalHypersurface)
```
Comment (martapanizzut) [inline src/TropicalGeometry/hypersurface.jl:264]: This function is not specific for tropical linear spaces. It should be combined with other functions for hypersurfaces which are currently work in progress. Maybe add a comment?

File: src/TropicalGeometry/semiring.jl [hunk 2]
```diff
@@ -435,6 +435,32 @@ function det(x::Matrix{Oscar.TropicalSemiringElem{T}}) where {T}
...
+################################################################################
+#
+#  Tropical Minors
+#
+################################################################################
+@doc Markdown.doc"""
+    tropical_minors(A::MatElem, k::Int)
+
+Return an array consisting of the k-minors of a tropical matrix A
+"""
```
Comment (martapanizzut) [inline src/TropicalGeometry/semiring.jl:449]: Add doctest.

File: src/TropicalGeometry/linear_space.jl [hunk 3]
```diff
@@ -32,29 +35,110 @@ export TropicalLinearSpace
...
 
+TropicalLinearSpace(plv::Vector{TropicalSemiringElem{typeof(min)}},rank::IntegerUnion, nElements::IntegerUnion) =
+TropicalLinearSpace_impl(plv, rank, nElements, min)
+
+TropicalLinearSpace(plv::Vector{TropicalSemiringElem{typeof(max)}},rank::IntegerUnion, nElements::IntegerUnion) =
+TropicalLinearSpace_impl(plv, rank, nElements, max)
+
+
+@doc Markdown.doc"""
+    TropicalLinearSpace()
```
Comment (martapanizzut) [inline src/TropicalGeometry/linear_space.jl:106]: Missing argument.

File: src/TropicalGeometry/linear_space.jl [hunk 4]
```diff
@@ -32,29 +35,110 @@ export TropicalLinearSpace
...
+    l = length(g)
+    id = [ v for v in identity_matrix(ZZ,n)]
+    A = [ [ coeff(p,Vector{Int}(v)) for v in [id[i,:] for i in 1:nrows(id)] ] for p in gens(I) ]
+    B = kernel_basis(matrix(QQ,A))
+    return TropicalLinearSpace(matrix(B),val)
 end
 
-
 @doc Markdown.doc"""
     TropicalLinearSpace(plv::Vector)
```
Comment (martapanizzut) [inline src/TropicalGeometry/linear_space.jl:69]: The signature is not complete.

**Comment by fieker** [comment]
On Wed, Mar 30, 2022 at 08:11:29AM -0700, panizzut.marta wrote:
> @martapanizzut commented on this pull request.
> 
> 
> 
> > @@ -253,3 +253,14 @@ function polynomial(TH::TropicalHypersurface{M,EMB}) where {M,EMB}
>      end
>      return get_attribute(TH,:tropical_polynomial)
>  end
> +export polynomial
> +
> +
> ***@***.*** Markdown.doc"""
> +    minimalPolynomialFromHypersurface(T::TropicalHypersurface)
> +
> +Returns the minimal polynomial with smallest possible coefficients of a hypersurface
> +"""
> +function minimalPolynomialFromHypersurface(T::TropicalHypersurface)
> 
> This function is not specific for tropical linear spaces. It should be combined with other functions for hypersurfaces which are currently work in progress. Maybe add a comment? 
> 
> > @@ -435,6 +435,32 @@ function det(x::Matrix{Oscar.TropicalSemiringElem{T}}) where {T}
>    return det(matrix(parent(x[1,1]),x))
>  end
>  
> +
Why is this not just called minpoly? the "FromHypersurface" is clear by
the argument and generically minpoly is the current generic name for
this...

**Comment by fieker** [comment]
A couple of minors:
 - why "Tropical"LinearSpace? I'd say the "Tropical" bit comes from the
   type of the ring already
 - indeed: should this (or could this) be a 
   free_module? Do we need a general "linear_space" rename for
   modules/vector spaces?

 - functions should be "snake_case", ie. in lower case with underscore
   and not "CamelCase" as per general julia convention.
   In fact, I'd suggest to provide also all constructors (at least
   if users are supposed to use them) in snake_case
   as for me, a user, it is impossible to predict if s.th. should be
   CamelCase (since it is a constructor) or snake_case (since it
   computes s.th.)

 - almost always, outside a constructor for a type, type names should not
   be part of function names: for examples, we don't do subgroup or
   submodule, we do sub (yes it is too short). This allows to write code
   in a lot more generic way. The type should be clear from the
   argument(s)
   There are of course also exceptions. But for example if I am already
   in the projective world, then I do not need to or want to decorate my
   function names with projective.

 - the code really does not have to be perfect 1st time round...but the
   sooner s.th. is mainstram (merged), the sooner people will actually
   see it, (ab)use it and provide feedback.

**Comment by fieker** [comment]
On Wed, Mar 30, 2022 at 01:06:48AM -0700, Lars Kastner wrote:
> Why was this merged? I specifically assigned to people to review it, it is their stuff. They have not done any reviewing yet. Also I am quite sure this was a draft.
> 
I had a discussion with one the reviewers and was informed that all is
as fine as far as known

**Pull request merged**